test/Microsoft.SqlTools.ServiceLayer.Test/QueryExecution/ExecutionPlanTests.cs

```diff
@@ -0,0 +1,251 @@
+            ResultSet planResultSet = b.ResultSets.First();
+            ExecutionPlan plan = planResultSet.GetExecutionPlan().Result;
+
+            // Then:
+            // ... I should get the execution plan back
+            Assert.Equal("xml", plan.Format);
+            Assert.Contains("Execution Plan", plan.Content);
+        }
+
+        [Fact]
+        public void ExecutionPlanInvalid()
+        {
+            // Setup:
+            // ... I have a batch that has been executed
+            Batch b = Common.GetBasicExecutedBatch();
+
+            // If:
+            // ... I have a result set and I ask for an execution plan that doesn't exist
+            ResultSet planResultSet = b.ResultSets.First();
+            ExecutionPlan plan = planResultSet.GetExecutionPlan().Result;
+
+            // Then:
+            // ... I should get a null value since it has no execution plan 
+            Assert.Null(plan);
+        }
+
+        #endregion
+
+        #region Batch Class Tests
+
+        [Fact]
+        public void BatchExecutionPlanValidTest()
+        {
+            // If I have an executed batch which has an execution plan 
+            Batch b = Common.GetExecutedBatchWithExecutionPlan();
+
+            // ... And I ask for a valid execution plan 
+            ExecutionPlan plan = b.GetExecutionPlan(0).Result;
+
+            // Then:
+            // ... I should get the execution plan back
+            Assert.Equal("xml", plan.Format);
+            Assert.Contains("Execution Plan", plan.Content);
+        }
+
+        [Fact]
+        public void BatchExecutionPlanInvalidTest()
+        {
+            // Setup:
+            // ... I have a batch that has been executed without an execution plan 
+            Batch b = Common.GetBasicExecutedBatch();
+
+            // If: 
+            // ... I ask for an invalid execution plan 
+            ExecutionPlan plan = b.GetExecutionPlan(0).Result;
+
+            // Then:
+            // ... I should get a null value since it has no execution plan 
+            Assert.Null(plan);
+        }
+
+        [Theory]
+        [InlineData(-1)]  // Invalid result set, too low
+        [InlineData(2)]   // Invalid result set, too high
+        public void BatchExecutionPlanInvalidParamsTest(int resultSetIndex)
+        {
+            // If I have an executed batch which has an execution plan 
+            Batch b = Common.GetExecutedBatchWithExecutionPlan();
+
+            // ... And I ask for an execution plan with an invalid result set index
+            // Then: 
+            // ... It should throw an exception
+            Assert.ThrowsAsync<ArgumentOutOfRangeException>(() => b.GetExecutionPlan(resultSetIndex)).Wait();
+        }
+
+        #endregion
+
+        #region Query Class Tests
+
+        [Theory]
+        [InlineData(-1)]  // Invalid batch, too low
+        [InlineData(2)]   // Invalid batch, too high
+        public void QueryExecutionPlanInvalidParamsTest(int batchIndex)
```

Comment:
It'd be my preference to try and isolate these sort of tests to the class they test. In this case, the batch class. So instead of creating a whole Query object, I'd just create the Batch and test from there. But, this works, too.